api/slim.Dockerfile

```
@@ -0,0 +1,13 @@
+FROM rust:alpine3.11 as builder
+LABEL stage=code-library-intermediate
+WORKDIR /usr/src/code-library-api
+COPY . .
+RUN apk add gcc musl-dev openssl-dev postgresql-dev
+ENV RUSTFLAGS="-C target-feature=-crt-static"
+RUN cargo install --path .
+
+FROM alpine:latest
+RUN apk add gcc libpq
+COPY --from=builder /usr/local/cargo/bin/api /usr/local/bin/api
+EXPOSE 8080/tcp
+CMD ["api"]
```

Have you tried running the image to launch the service locally? I have a bit different set up and the problem I have is that it doesn't start the `api` service, but rather launches another mircroservice (book, borrow etc), despite the fact that the entrypoint is defined as `api`. 

This is the set up I tried:
```suggestion
FROM rust
RUN USER=root cargo new --bin api

WORKDIR /api
COPY ./Cargo.toml ./Cargo.toml

RUN cargo build
RUN rm src/*.rs

COPY ./src ./src
COPY ./diesel.toml ./diesel.toml

RUN rm ./target/debug/deps/api*
RUN cargo build --release

CMD ["./target/debug/api"]
EXPOSE 8000

```
(It assumes that the dependencies won't rebuild when used locally)

---

The Dockerfile I created works without any problem. It's the one which I currently use for the deployment in development. For your suggested code I get errors.

---

that's weird, I get the same behavior with your version as well :/  how do you launch the image on aws or locally?

---

the suggested version assumes re-use of the image because of the `cargo.toml` separate building, I'm not 100% sure if it's something we need to use with CI/CD as circleci still does internal caching and we always use the `cargo.toml` with the latest changes anyway to start build-push process, but it's definitely helpful locally - for testing for example

---

I yesterday corrected the copy path, because it didn't match the new directory structure. Maybe this was the problem. I build the image with `docker build -t code-library-notification:v0.1 .` and launch it with `docker run -t -P -e "DATABASE_URL=<db-url>" -e "OAUTH_CLIENT_IDENTIFIER=<client-identifier>" -e "OAUTH_CLIENT_SECRET=<client-secret>" -e "OAUTH_CLIENT_REDIRECT=<client-redirect>" -e "JWT_SECRET=<jwt-secret>" localhost/code-library-notification:v0.1`. This is the same on server and on my local machine.

---

Got it, thanks! let's use the version you created then if it already works well on aws.